drivers/include/net/netdev/ieee802154.h

```diff
@@ -115,14 +121,65 @@ typedef struct {
     uint8_t page;                           /**< channel page */
     uint16_t flags;                         /**< flags as defined above */
     int16_t txpower;                        /**< tx power in dBm */
+    const netdev_ieee802154_ops_t *ops;     /**< IEEE802.15.4 device ops. Must be set */
     /** @} */
 } netdev_ieee802154_t;
 
+/**
+ * @brief Operations of an IEEEE802.15.4 device
+ */
+struct netdev_ieee802154_ops {
+    /**
+     * @brief Set IEEE addresses to the hardware accelerator of an
+     * network device
+     *
+     * @pre (short_addr != NULL), (ext_addr != NULL)
+     * @note Set to NULL if the device doesn't support address filtering.
+     *
+     * @param[in] dev           network device descriptor
+     * @param[in] short_addr    short address to be set
+     * @param[in] ext_addr      extended address to be set
+     * @param[in] panid         panid to be set
+     */
+    void (*set_hw_addr)(netdev_ieee802154_t *dev, network_uint16_t *short_addr,
+                        eui64_t *ext_addr, uint16_t panid);
+};
+
 /**
  * @brief   Received packet status information for IEEE 802.15.4 radios
  */
 typedef struct netdev_radio_rx_info netdev_ieee802154_rx_info_t;
 
+/**
+ * @brief Check whether a @ref netdev_ieee802154_t device has hardware address
+ *        filter support
+ *
+ * @param[in] dev       network device descriptor
+ *
+ * @return              true if the device supports address filtering
+ */
+static inline bool netdev_ieee802154_has_addr_filter(netdev_ieee802154_t *dev)
+{
+    return dev->ops->set_hw_addr != NULL;
+}
+
+/**
+ * @brief Configure hardware address filter in a given network device
+ *
+ * @pre   @ref dev MUST have hardware address filter! Check with
+ *        @ref netdev_ieee802154_has_addr_filter
+ *
+ * @param[i] dev        network device descriptor
+ * @param[i] short_addr IEEE802.15.4 short address
+ * @param[i] ext_addr   IEEE802.15.4 extended address
+ * @param[i] panid      IEEE802.15.4 PAN-ID
+ */
+static inline void netdev_ieee802154_set_addr_filter(netdev_ieee802154_t *dev, network_uint16_t *short_addr, eui64_t *ext_addr, uint16_t panid)
+{
+    assert(dev->ops->set_hw_addr);
+    dev->ops->set_hw_addr(dev, short_addr, ext_addr, panid);
```

Comment:
Is this function ever called without `netdev_ieee802154_has_addr_filter` called? Why not just do

```C
if (dev->ops->set_hw_addr != NULL) {
    dev->ops->set_hw_addr(dev, short_addr, ext_addr, panid);
}
```

Reply:
> Is this function ever called without netdev_ieee802154_has_addr_filter called?

I'm not sure to be honest. I just wanted to make it explicit to the user (so it's not a NOP)

Reply:
> I'm not sure to be honest.

Well as you would otherwise run into a NULL pointer dereference, you *have* to check at some point.

> I just wanted to make it explicit to the user (so it's not a NOP)

The user would be the `netif` module, right?

Regarding NOP: it wouldn't be one with a return value.

Reply:
> Well as you would otherwise run into a NULL pointer dereference, you have to check at some point.

Unless you know that the device has address filters (e.g you asked before). But maybe is an extreme use case.

> The user would be the netif module, right?

The direct user now is `netdev_ieeee802154_set`, and this is the one used by `netif`.
So, in practice only lower layer developers would call this operations.
A user of raw radios is a usually a user of a PHY layer with (sub)MAC (CSMA, retransmissions, etc), so they are also excluded unless they really want to have control of the hardware address filter

> Regarding NOP: it wouldn't be one with a return value.

Considering a user would never call this directly, the desired operation in this case is: "if there's hardware acceleration available, set it to the following parameters".
There's also a second operation needed for software address filtering: "If the device doesn't support hardware address filtering, then trigger the software address filtering mechanism".
So, with this 2 new functions it should be possible to explicitly implement these features.

Reply:
Mh.. ok, but then I think the `netdev_ieee802154_set_addr_filter()` function is "overkill". `netdev_ieee802154_has_addr_filter()` is nice in case the MAC developer really doesn't know what they are doing, but the only advantage of `netdev_ieee802154_set_addr_filter()` is the sanity check. I would just spell it out in the using code:

```C
if (netdev_ieee802154_has_addr_filter(dev)) {
    dev->ops->set_hw_addr(dev, short_addr, ext_addr, panid);
}

Reply:
I get the fact that `netdev_ieee802154_set_addr_filter` is only a wrapper in the current state.
However, I wanted to do it this way to avoid exposing struct members (if the ops structure changes for whatever reason, it doesn't break the API).
Also, it's exposed in the documentation at the same level as other `netdev_ieee802154_xxx` functions.

Since it's `static inline`, it doesn't increase binary size

Reply:
Ok, given that this is one of the first steps towards a major rework, I agree.